File: topicos/dynamic-programming/group-1/binary_search/binary_search.ts
 export function binarySearch(nums: number[], x: number): number {
    let left = 0;
    let right = nums.length-1;

    let mid = Math.round(right/2);

    while (left <= right) {
        let mid = Math.round(left+(right-right) >>> 1);

        if (nums[mid] == x) return mid;
        
        if (nums[mid] > x) {
            left = mid+1;
        } else {
            right = mid-1;
        }
    }
}

File: topicos/dynamic-programming/group-1/binary_search/binary_search.rs
use std::cmp::Ordering;

pub fn binary_search<T: Ord>(arr: &[T], target: &T) -> Option<usize> {
    if arr.is_empty() {
        return None;
    }

    let mut left = 0;
    let mut right = arr.len() - 1;

    while left <= right {
        let mid = left + (right - left) / 2;

        match arr[mid].cmp(target) {
            Ordering::Equal => return Some(mid),
            Ordering::Less => left = mid + 1,
            Ordering::Greater => {
                if mid == 0 {
                    break;
                }

                right = mid - 1;
            },
        }
    }

    None
}